src/metabase/models/segment.clj

```diff
@@ -86,37 +86,37 @@
         (log/errorf e "Error calculating Segment description: %s" (ex-message e))
         nil))))
 
-(mu/defn ^:private warmed-metadata-provider :- lib.metadata/MetadataProvider
+(mu/defn ^:private warmed-metadata-provider :- ::lib.schema.metadata/metadata-provider
   [database-id :- ::lib.schema.id/database
    segments    :- [:maybe [:sequential (ms/InstanceOf :model/Segment)]]]
   (let [metadata-provider (doto (lib.metadata.jvm/application-database-metadata-provider database-id)
                             (lib.metadata.protocols/store-metadatas!
-                             :metadata/segment
                              (map #(lib.metadata.jvm/instance->metadata % :metadata/segment)
                                   segments)))
         field-ids         (mbql.u/referenced-field-ids (map :definition segments))
-        fields            (lib.metadata.protocols/bulk-metadata metadata-provider :metadata/column field-ids)
+        fields            (lib.metadata.protocols/metadatas metadata-provider :metadata/column field-ids)
         table-ids         (into #{}
                                 cat
                                 [(map :table-id fields)
                                  (map :table_id segments)])]
     ;; this is done for side effects
-    (lib.metadata.protocols/bulk-metadata metadata-provider :metadata/table table-ids)
+    (when (lib.metadata.protocols/cached-metadata-provider? metadata-provider)
+      (lib.metadata.protocols/metadatas metadata-provider :metadata/table table-ids))
```

Comment:
This is a bit ugly to see outside of `metabase.lib.metadata.*`.

How about `(lib.metadata/warm-cache metadata-provider :metadata/table table-ids)` (or in `lib.metadata.protocols` if that's a better fit) and let that understand about `cached-metadata-provider?` etc. and do nothing if there's no cache.

Reply:
Yeah that's a good point

Reply:
Implemented this